test/new_relic/agent/log_event_aggregator_test.rb

```diff
@@ -521,5 +527,164 @@ def test_records_log_events_not_within_default_severities
         assert_equal(log_message, events.first.last['message'])
       end
     end
+
+    def test_record_json_sets_severity_when_given_level
+      @aggregator.record_logstasher_event({'level' => :warn, 'message' => 'yikes!'})
+      _, events = @aggregator.harvest!
+
+      assert_equal :warn, events[0][1]['level']
+      assert_metrics_recorded([
+        'Logging/lines/warn'
+      ])
+    end
+
+    def test_record_json_sets_severity_unknown_when_no_level
+      @aggregator.record_logstasher_event({'message' => 'hello there'})
+      _, events = @aggregator.harvest!
+
+      assert_equal 'UNKNOWN', events[0][1]['level']
+      assert_metrics_recorded([
+        'Logging/lines/UNKNOWN'
+      ])
+    end
+
+    def test_record_json_does_not_record_if_message_is_nil
+      @aggregator.record_logstasher_event({'level' => :info, 'message' => nil})
+      _, events = @aggregator.harvest!
+
+      assert_empty events
+    end
+
+    def test_record_json_does_not_record_if_message_empty_string
+      @aggregator.record_logstasher_event({'level' => :info, 'message' => ''})
+      _, events = @aggregator.harvest!
+
+      assert_empty events
+    end
+
+    def test_record_json_returns_message_when_avaliable
+      @aggregator.record_logstasher_event({'level' => :warn, 'message' => 'A trex is near'})
+      _, events = @aggregator.harvest!
+
+      assert_equal 'A trex is near', events[0][1]['message']
+    end
+
+    def test_record_json_returns_empty_string_when_unavaliable
+      @aggregator.record_logstasher_event({'level' => :info})
+      _, events = @aggregator.harvest!
+
+      assert_equal '', events[0][1]['message']
+    end
+
+    def test_add_json_event_attributes_records_attributes
+      @aggregator.record_logstasher_event({'source' => '127.0.0.1', 'tags' => ['log']})
+      _, events = @aggregator.harvest!
+
+      assert_includes(events[0][1]['attributes'], 'source')
+      assert_includes(events[0][1]['attributes'], 'tags')
+    end
+
+    def test_add_json_event_attributes_deletes_already_recorded_attributes
+      @aggregator.record_logstasher_event({'message' => 'bye', 'level' => :info, '@timestamp' => 'now'})
+      _, events = @aggregator.harvest!
+
+      refute_includes(events[0][1]['attributes'], 'message')
+      refute_includes(events[0][1]['attributes'], 'level')
+      refute_includes(events[0][1]['attributes'], '@timestamp')
+    end
```

Comment:
What do you think about adding one attribute that wouldn't fall under the "already recorded" category? This might help us make sure attributes aren't being deleted if they're present alongside the "already recorded" attributes? 

Also, could you add some expectations to verify these attributes are recorded elsewhere in the event?

Reply:
Great idea! https://github.com/newrelic/newrelic-ruby-agent/pull/2559/commits/84c64b3374ec74d65b9e3fe130128c648d8cf670